reactive/postfix.py

```diff
@@ -0,0 +1,248 @@
+        smtpd_recipient_restrictions.append(
+            "check_sender_access hash:/etc/postfix/restricted_senders"
+        )
+    smtpd_recipient_restrictions.extend(charm_state.additional_smtpd_recipient_restrictions)
+
+    if charm_state.enable_spf:
+        smtpd_recipient_restrictions.append("check_policy_service unix:private/policyd-spf")
+
+    return smtpd_recipient_restrictions
+
+
+def construct_postfix_config_file_content(  # pylint: disable=too-many-arguments
+    *,
+    charm_state: "State",
+    tls_dh_params_path: str,
+    tls_cert_path: str,
+    tls_key_path: str,
+    tls_cert_key_path: str,
+    fqdn: str,
+    hostname: str,
+    milters: str,
+    template_path: str,
+) -> str:
+    """Prepare the context and renders the Postfix configuration files."""
+    context = {
+        "JUJU_HEADER": utils.JUJU_HEADER,
+        "fqdn": fqdn,
+        "hostname": hostname,
+        "connection_limit": charm_state.connection_limit,
+        "enable_rate_limits": charm_state.enable_rate_limits,
+        "enable_sender_login_map": bool(charm_state.sender_login_maps),
+        "enable_smtp_auth": charm_state.enable_smtp_auth,
+        "enable_spf": charm_state.enable_spf,
+        "enable_tls_policy_map": bool(charm_state.tls_policy_maps),
+        "header_checks": bool(charm_state.header_checks),
+        "milter": milters,
+        "mynetworks": ",".join(charm_state.allowed_relay_networks),
+        "relayhost": charm_state.relay_host,
+        "relay_domains": " ".join(charm_state.relay_domains),
+        "relay_recipient_maps": bool(charm_state.relay_recipient_maps),
+        "restrict_recipients": bool(charm_state.restrict_recipients),
+        "smtp_header_checks": bool(charm_state.smtp_header_checks),
+        "smtpd_recipient_restrictions": ", ".join(_smtpd_recipient_restrictions(charm_state)),
+        "smtpd_relay_restrictions": ", ".join(_smtpd_relay_restrictions(charm_state)),
+        "smtpd_sender_restrictions": ", ".join(_smtpd_sender_restrictions(charm_state)),
+        "tls_cert_key": tls_cert_key_path,
+        "tls_cert": tls_cert_path,
+        "tls_key": tls_key_path,
+        "tls_ciphers": charm_state.tls_ciphers.value if charm_state.tls_ciphers else None,
+        "tls_dh_params": tls_dh_params_path,
+        "tls_exclude_ciphers": ", ".join(charm_state.tls_exclude_ciphers),
+        "tls_protocols": " ".join(charm_state.tls_protocols),
+        "tls_security_level": (
+            charm_state.tls_security_level.value if charm_state.tls_security_level else None
+        ),
+        "transport_maps": bool(charm_state.transport_maps),
+        "virtual_alias_domains": " ".join(charm_state.virtual_alias_domains),
+        "virtual_alias_maps": bool(charm_state.virtual_alias_maps),
+        "virtual_alias_maps_type": charm_state.virtual_alias_maps_type.value,
+    }
+
+    return utils.render_jinja2_template(context, template_path)
+
+
+def _create_update_map(content: str, postmap: str) -> bool:
```

Comment:
Can you refactor this so that the interactions with the filesystem and the `subprocess.call` are extracted from the method? This is interaction with the container, so I'd be better to have it moved into the charm.py. We can think about refactoring it later because this will probably end up in several duplicates